pkg/collector/process.go

```diff
@@ -361,47 +375,94 @@ func (cp *CollectingProcess) decodeDataSet(dataBuffer *bytes.Buffer, obsDomainID
 func (cp *CollectingProcess) addTemplate(obsDomainID uint32, templateID uint16, elementsWithValue []entities.InfoElementWithValue) {
 	cp.mutex.Lock()
 	defer cp.mutex.Unlock()
-	if _, exists := cp.templatesMap[obsDomainID]; !exists {
-		cp.templatesMap[obsDomainID] = make(map[uint16][]*entities.InfoElement)
+	if _, ok := cp.templatesMap[obsDomainID]; !ok {
+		cp.templatesMap[obsDomainID] = make(map[uint16]*template)
 	}
 	elements := make([]*entities.InfoElement, 0)
 	for _, elementWithValue := range elementsWithValue {
 		elements = append(elements, elementWithValue.GetInfoElement())
 	}
-	cp.templatesMap[obsDomainID][templateID] = elements
-	// template lifetime management
+	tpl, ok := cp.templatesMap[obsDomainID][templateID]
+	if !ok {
+		tpl = &template{}
+		cp.templatesMap[obsDomainID][templateID] = tpl
+	}
+	tpl.ies = elements
+	klog.V(4).InfoS("Added template to template map", "obsDomainID", obsDomainID, "templateID", templateID)
+	// Template lifetime management for UDP.
 	if cp.protocol != "udp" {
 		return
 	}
-	// Handle udp template expiration
-	go func() {
-		ticker := time.NewTicker(cp.templateTTL)
-		defer ticker.Stop()
-		select {
-		case <-ticker.C:
+	tpl.expiryTime = cp.clock.Now().Add(cp.templateTTL)
+	if tpl.expiryTimer == nil {
+		tpl.expiryTimer = cp.clock.AfterFunc(cp.templateTTL, func() {
 			klog.Infof("Template with id %d, and obsDomainID %d is expired.", templateID, obsDomainID)
-			cp.deleteTemplate(obsDomainID, templateID)
-			break
+			now := cp.clock.Now()
+			// From the Go documentation:
+			//   For a func-based timer created with AfterFunc(d, f), Reset either
+			//   reschedules when f will run, in which case Reset returns true, or
+			//   schedules f to run again, in which case it returns false. When Reset
+			//   returns false, Reset neither waits for the prior f to complete before
+			//   returning nor does it guarantee that the subsequent goroutine running f
+			//   does not run concurrently with the prior one. If the caller needs to
+			//   know whether the prior execution of f is completed, it must coordinate
+			//   with f explicitly.
+			// In our case, when f executes, we have to verify that the record is indeed
+			// scheduled for deletion by checking expiryTime. We cannot just
+			// automatically delete the template.
+
+			// No reason to try to stop the timer in this case, even though it would be
```

Comment:
Trying to follow the logic, is it because the expiryTimer is already nil, so there's no point to set stopTimer to true? Because I was thinking about L444, wondering if there's a case when stopTimer is true but expiryTimer is nil.

Reply:
this code executes as part of the function invoked by the timer. Broadly speaking, if the function has been invoked by the timer, then the timer has already fired and there is no need to stop it. There are some edge cases with `AfterFunc` which are described in the documentation snippet above.
Essentially there are 2 cases:
- the timer fires because we need to expire the template. This is the "normal" case. The timer has fired and there is no need to stop it.
- the timer fires, but there is a concurrent call to `addTemplate` (and to `AfterFunc`). This is the edge case described in the doc. We rely on the condition below (`!tpl.expiryTime.After(now)`) to determine whether we should just ignore this firing of the timer. The synchronization provided by `cp.mutex` is very important here. We will wait until the call to `addTemplate` returns and releases the lock. If the condition is false (which will be the case), we will not call `expiryTimer.Stop()` in `deleteTemplateWithConds` (even if `stopTimer` was set to true), as the template entry has been updated and the current run of the timer function should be ignored and be a no-op.

Calling `expiryTimer.Stop()` even if the timer has already been stopped is harmless. But it doesn't work well with the fake clock unfortunately. If the fake clock implementation is fixed, I may remove `stopTimer`.

The `nil` check at line L444 is more for defensive programming and for the TCP case.